crates/astria-merkle/src/lib.rs

```diff
@@ -0,0 +1,718 @@
+        let Self {
+            audit_path,
+            leaf_index,
+            tree_size,
+        } = self;
+        let mut i = leaf_index_to_tree_index(*leaf_index);
+        let mut acc = leaf_hash;
+        for sibling in audit_path.chunks(32) {
+            let parent = complete_parent(i, *tree_size);
+            if parent > i {
+                acc = combine(&acc, sibling);
+            } else {
+                acc = combine(sibling, &acc);
+            }
+            i = parent;
+        }
+        acc
+    }
+
+    /// Walks the audit path to reconstruct the root hash starting from a node.
+    ///
+    /// Use this method if the starting point is a node (like a leaf), which will
+    /// need to be hashed prior to combining with its sibling.
+    ///
+    /// # Examples
+    /// ```
+    /// let mut tree = astria_merkle::Tree::new();
+    /// tree.push(&[1; 32]);
+    /// tree.push(&[2; 32]);
+    /// tree.push(&[3; 32]);
+    /// tree.push(&[4; 32]);
+    /// tree.push(&[5; 32]);
+    /// tree.push(&[6; 32]);
+    /// tree.push(&[7; 32]);
+    /// tree.push(&[8; 32]);
+    /// let root = tree.root();
+    /// let proof = tree
+    ///     .construct_proof(6)
+    ///     .expect("leaf 7 must be inside the tree");
+    /// let reconstructed_root = proof.reconstruct_root_with_leaf(&[7_u8; 32]);
+    /// assert_eq!(root, reconstructed_root);
+    /// ```
+    #[must_use]
+    pub fn reconstruct_root_with_leaf(&self, node: &[u8]) -> [u8; 32] {
+        self.reconstruct_root_with_leaf_hash(hash_leaf(node))
+    }
+}
+
+/// A low-level API to construct a leaf-hash ad-hoc without needing to allocate it.
+///
+/// [`LeafBuilder`] can only be created by using the [`Tree::build_push_leaf`] method.
+/// The leaf is built and the tree updated once `LeafBuilder` goes out of scope.
+///
+/// No two leaf builders can exist at the same time because a builder holds a mutable
+/// reference to its tree during its lifetime.
+///
+/// See [`Tree::build_leaf`] for usage.
+pub struct LeafBuilder<'a> {
+    tree: &'a mut Tree,
+    hasher: Option<Sha256>,
+}
+
+impl<'a> LeafBuilder<'a> {
+    /// Takes ownership of the builder, dropping it.
+    ///
+    /// This method causes the leaf builder to go out of scope, causing it
+    /// to be dropped and triggering an update of the merkle tree. Use this
+    /// if the builder is explicitly bound.
+    ///
+    /// Usually calling this is not necessary because writes to the builder can
+    /// be chained, but this is still useful if some other operation has to happen
+    /// between writes, requiring the builder to be found.
+    ///
+    /// # Examples
+    /// ```
+    /// let mut tree_1 = astria_merkle::Tree::new();
+    /// let mut builder = tree_1.build_leaf();
+    /// builder.write(b"hello");
+    /// builder.write(b"world");
+    /// builder.finish();
+    ///
+    /// let mut tree_2 = astria_merkle::Tree::new();
+    /// tree_2.build_leaf().write(b"hello").write(b"world");
```

Comment:
```suggestion
    /// tree_2.build_leaf().write(b"helloworld");
```

the two trees in the example are built the same way, is this understanding correct?

Reply:
also `finish` isn't called for `tree_2`

Reply:
This is just for demonstration purposes to show the `build_leaf` API. :-) This is used to avoid extra allocations.

Maybe I should choose another example?

Reply:
> also `finish` isn't called for `tree_2`

Just saw that after commenting: `finish` need not be called because the tree is updated on drop. See the comment on this method: 
```
    /// Usually calling this is not necessary because writes to the builder can
    /// be chained, but this is still useful if some other operation has to happen
    /// between writes, requiring the builder to be found.
```

Reply:
ah okay, yeah i just meant it's a bit weird that `tree_1` and `tree_2` in the examples are built the same :p

Reply:
I have added a 3rd tree to that example that demonstrates that the two above methods are equivalent to calling `push(b"helloworld");`